i2c/ina228_i2c/ina228_i2c.c

```diff
@@ -0,0 +1,119 @@
+#include <stdio.h>
+#include "pico/stdlib.h"
+#include "hardware/i2c.h"
+#include <math.h>
+
+// LED GPIOs
+#define EXT_LED_GPIO 2
+
+// INA228 integrated shunt resistance
+#define R_SHUNT 0.015
+
+// Max expected current through shunt resistor
+#define MAX_EXPECTED_CURRENT 0.1
+#define I2C_ADDR 0x40
+
+// ina228 registers (see datasheet)
+uint8_t VSHUNT_REG = 0x04;
+uint8_t VBUS_REG = 0x05;
+uint8_t DIETEMP_REG = 0x06;
+uint8_t CURRENT_REG = 0x07;
+uint8_t POWER_REG = 0x08;
+uint8_t ENERGY_REG = 0x09;
+uint8_t CHARGE_REG = 0x0A;
+
+uint8_t ADC_CONFIG_REG = 0x01;
+uint8_t SHUNT_CAL_REG = 0x02;
+
+// conversion factors (see datasheet)
+// note: VSHUNT_FACTOR assumes ADCRANGE = 0 
+const double VSHUNT_FACTOR = 312.5 * 1e-9;
+const double VBUS_FACTOR = 195.3125 * 1e-6;
+const double DIETEMP_FACTOR  = 7.8125 * 1e-3;
+const double CURRENT_FACTOR = MAX_EXPECTED_CURRENT / (1<<19);
+const double POWER_FACTOR = 3.2 * CURRENT_FACTOR ;
+const double ENERGY_FACTOR  = 16 * 3.2 * CURRENT_FACTOR;
+const double CHARGE_FACTOR = CURRENT_FACTOR;
+
+const uint16_t SHUNT_CAL = 13107.2 * 1e6 * CURRENT_FACTOR * R_SHUNT;
+
+float vshunt, vbus, dietemp, current, power, energy, charge;
+
+static void ina228_init() {
+    // Write to shunt register 
+    uint8_t shunt_msb = (SHUNT_CAL >> 8) & 0xFF;
+    uint8_t shunt_lsb = SHUNT_CAL & 0xff;
+    uint8_t shunt_buf[3] = {SHUNT_CAL_REG, shunt_msb, shunt_lsb};
+    i2c_write_blocking(i2c_default, I2C_ADDR, shunt_buf, 3, false);
+
+    // Write to ADC config register (need to enable continuous mode to access accumilation variables energy and charge)
+    uint8_t adc_msb = 0xF0;
+    uint8_t adc_lsb = 0x00;
+    uint8_t adc_buf[3] = {ADC_CONFIG_REG, adc_msb, adc_lsb};
+    i2c_write_blocking(i2c_default, I2C_ADDR, adc_buf, 3, false);
+}
+
+static void ina228_read(float *vshunt, float *vbus, float *dietemp, float *current, float *power, float *energy, float *charge) {
+    // Buffers for writing register measurments to
+    // Some registers are 2 bytes, some are 3 bytes, the accumilation registers are 5 bytes
+    uint8_t buffer_2[2];
+    uint8_t buffer_3[3];
+    uint8_t buffer_5[5];
+
+    // For some of the registers the 4 least significant bits are reserved and always read zero, which is why we shift right 12, 4, then left 4 for vshunt, vbus and current
```

Comment:
I'm not sure that you need so much detail here, it took me a while to understand what this comment was saying! Perhaps something like this is sufficient?
```suggestion
    // For the vshunt, vbus and current registers the 4 least significant bits are reserved and always read zero, so use bit-shifts to just extract the actual data
```